File: 722.删除注释.go
package leetcode

/*
 * @lc app=leetcode.cn id=722 lang=golang
 *
 * [722] 删除注释
 */

// @lc code=start
func removeComments(source []string) (res []string) {
	var inBlock bool
	var newLine []byte
	for _, line := range source {
		n := len(line)
		for i := 0; i < n; i++ {
			if inBlock {
				if i+1 < n && line[i] == '*' && line[i+1] == '/' {
					inBlock = false
					i++
				}
			} else {
				if i+1 < n && line[i] == '/' && line[i+1] == '*' {
					inBlock = true
					i++
				} else if i+1 < n && line[i] == '/' && line[i+1] == '/' {
					break
				} else {
					newLine = append(newLine, line[i])
				}
			}
		}
		if !inBlock && len(newLine) > 0 {
			res = append(res, string(newLine))
			newLine = nil
		}
	}
	return
}

// @lc code=end

File: 722.删除注释.c
#include<stdbool.h>
/*
 * @lc app=leetcode.cn id=722 lang=c
 *
 * [722] 删除注释
 */

 // @lc code=start
 /**
  * Note: The returned array must be malloced, assume caller calls free().
  */
#define MAX_LINE_LEN 80

char** removeComments(char** source, int sourceSize, int* returnSize) {
    char** res = (char**)calloc(sourceSize, sizeof(char*));
    char new_line[sourceSize * MAX_LINE_LEN + 1];
    int pos = 0, new_line_pos = 0;
    bool in_block = 0;
    for (int j = 0; j < sourceSize; j++) {
        char* line = source[j];
        int line_size = strlen(line);
        for (int i = 0; i < line_size; i++) {
            if (in_block) {
                if (i + 1 < line_size && line[i] == '*' && line[i + 1] == '/') {
                    in_block = false;
                    i++;
                }
            }
            else {
                if (i + 1 < line_size && line[i] == '/' && line[i + 1] == '*') {
                    in_block = true;
                    i++;
                }
                else if (i + 1 < line_size && line[i] == '/' && line[i + 1] == '/') {
                    break;
                }
                else {
                    new_line[new_line_pos++] = line[i];
                }
            }

        }
        if (!in_block && new_line_pos > 0) {
            new_line[new_line_pos] = '\0';
            res[pos] = (char*)calloc(new_line_pos + 1, sizeof(char));
            strcpy(res[pos], new_line);
            pos++;
            new_line_pos = 0;
        }
    }
    *returnSize = pos;
    return res;

}
// @lc code=end

